applications/nrf_desktop/src/hw_interface/selector_hw.c

```diff
@@ -0,0 +1,352 @@
+
+#define DEFAULT_STATE UCHAR_MAX
+#define DEBOUNCE_DELAY K_MSEC(5)
+
+
+enum state {
+	STATE_DISABLED,
+	STATE_ACTIVE,
+	STATE_OFF,
+	STATE_ERROR,
+};
+
+struct selector {
+	const struct selector_config *config;
+	struct gpio_callback gpio_cb[ARRAY_SIZE(port_map)];
+	struct k_delayed_work work;
+	u8_t state;
+	u8_t read_port;
+	u32_t read_pins_mask;
+};
+
+static struct device *gpio_dev[ARRAY_SIZE(port_map)];
+static struct selector selectors[ARRAY_SIZE(selector_configs)];
+static enum state state;
+
+
+static void selector_event_send(const struct selector *selector)
+{
+	BUILD_ASSERT_MSG(ARRAY_SIZE(selector_configs) <= UCHAR_MAX,
+			 "Maximum number of selectors is 255");
+	__ASSERT_NO_MSG((selector->state < selector->config->pins_size) ||
+			(selector->state == DEFAULT_STATE));
+
+	struct selector_event *event = new_selector_event();
+
+	event->selector_id = selector->config->id;
+	event->state = selector->state;
+
+	EVENT_SUBMIT(event);
+}
+
+static int read_state(struct selector *selector)
+{
+	u32_t value = 0;
+	const struct gpio_pin *sel_pins = selector->config->pins;
+
+	for (size_t i = 0; i < selector->config->pins_size; i++) {
+		int err = gpio_pin_read(gpio_dev[sel_pins[i].port],
+					sel_pins[i].pin,
+					&value);
+
+		if (err) {
+			LOG_ERR("Cannot read value from port:%" PRIu8
+				" pin: %" PRIu8, sel_pins[i].port,
+				sel_pins[i].pin);
+			return err;
+		}
+
+		if (value) {
+			selector->state = i;
+			break;
+		}
+	}
+
+	if (!value) {
+		selector->state = DEFAULT_STATE;
+		LOG_WRN("Selector %" PRIu8 " state is unknown",
+			selector->config->id);
+	}
+	selector_event_send(selector);
+
+	return 0;
+}
+
+static int enable_interrupts_lock(struct selector *selector)
+{
+	const struct gpio_pin *sel_pins = selector->config->pins;
+	int key = irq_lock();
+
+	for (size_t i = 0; i < selector->config->pins_size; i++) {
+		if (i == selector->state) {
+			continue;
+		}
+
+		int err = gpio_pin_enable_callback(gpio_dev[sel_pins[i].port],
+						   sel_pins[i].pin);
+		if (err) {
+			LOG_ERR("Cannot enable callback (err %d)", err);
+			irq_unlock(key);
+			return err;
+		}
+	}
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int disable_interrupts_nolock(struct selector *selector)
+{
+	const struct gpio_pin *sel_pins = selector->config->pins;
+
+	for (size_t i = 0; i < selector->config->pins_size; i++) {
+		int err = gpio_pin_disable_callback(gpio_dev[sel_pins[i].port],
+						    sel_pins[i].pin);
+
+		if (err) {
+			LOG_ERR("Cannot disable callback (err %d)", err);
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+static void selector_isr(struct device *dev, struct gpio_callback *cb,
+			 u32_t pins_mask)
+{
+	u8_t port = ARRAY_SIZE(port_map);
+
+	for (size_t i = 0; i < ARRAY_SIZE(gpio_dev); i++) {
+		if (gpio_dev[i] == dev) {
+			port = i;
+			break;
+		}
+	}
+	__ASSERT_NO_MSG(port < ARRAY_SIZE(port_map));
+
+	struct selector *sel;
+
+	sel = CONTAINER_OF((u8_t *)cb - port * sizeof(struct gpio_callback),
```

Comment:
`cb - port * sizeof(struct gpio_callback)`
what about padding?

Reply:
Is it possible that padding will not be taken into account by `sizeof(struct gpio_callback)` ?

Reply:
on the second thought no, please keep it but instead of sizeof(struct xxx) use sizeof(sel->gpio_cb[0])

Reply:
ok